lib/fileset/default.nix

```diff
@@ -151,7 +153,12 @@ If a directory does not recursively contain any file, it is omitted from the sto
       sourceFilter = _toSourceFilter fileset;
     in
     if ! isPath root then
-      if isStringLike root then
+      if root ? _isLibCleanSourceWith then
+        throw ''
+          lib.fileset.toSource: `root` is a `lib.sources`-based value, but it should be a path instead.
+              To use a `lib.sources`-based value, convert it to a file set using `lib.fileset.fromSource` and pass it as `fileset`.
+              Note that this only works for sources created from paths.''
```

Comment:
Any reason not to just call it for them?

Reply:
I'm not 100% convinced it's a good idea, just minor ideas like
- The operation this function does is somewhat expensive, so it being explicit makes people aware of that. This is different from the implicit path coercion, which is very cheap
- Do we want to deprecate this at some point? Probably not
- Source filters can be impure by changing behavior based on the project directory, does this impurity leak into file sets? Probably not a problem
- Is it a problem to have a lack of symmetry regarding `toSource`/`fromSource`? Probably not, these aren't symmetric, fromSource discard's the `root` and `toSource` has to add it
- Would it be a problem to have cases like `toSource { root = ./.; fileset = toSource { root = ./.; fileset = toSource { ...`? Probably also no

Furthermore this being an error for now means that we could still do this later.